raiden-ts/src/transfers/epics/locked.ts

```diff
@@ -649,81 +652,172 @@ function receiveTransferExpired(
+  const confirmation = action.payload.message;
+  const tokenNetwork = confirmation.token_network_address;
+  const partner = action.meta.address;
   const meta = {
+    direction: Direction.SENT, // received confirmation is for sent withdraw request
     tokenNetwork,
     partner,
-    totalWithdraw: request.total_withdraw,
-    expiration: request.expiration.toNumber(),
+    totalWithdraw: confirmation.total_withdraw,
+    expiration: confirmation.expiration.toNumber(),
   };
-  // empty/undefined confirmation$ or signing errors are non-fatal, we should still accept
-  // WithdrawRequest state change (which increases partner's nonce) and wait for WithdrawExpired
-  return merge(
-    of(withdrawReceive.request({ message: request }, meta)),
-    confirmation$.pipe(
-      filter(isntNil),
-      map((signed) => {
-        assert(request.expiration.gt(state.blockNumber), 'request expired while signing');
-        return withdrawReceive.success({ message: signed }, meta);
-      }),
-      catchError(
-        (err) => (log.warn('error while signing WithdrawConfirmation, ignoring', err), EMPTY),
-      ),
-    ),
+  return state$.pipe(
+    first(),
+    map((state) => {
+      const channel = state.channels[channelKey({ tokenNetwork, partner })];
+      assert(channel?.state === ChannelState.open, 'channel not open');
+
+      assert(confirmation.participant === address, 'participant mismatch');
+      assert(confirmation.chain_id.eq(state.chainId), 'chainId mismatch');
+      assert(confirmation.channel_identifier.eq(channel.id), 'channelId mismatch');
+
+      // don't validate req here, to always update nonce, but do on tx send
+      return withdrawMessage.success({ message: confirmation }, meta);
+    }),
+    catchError((err) => of(withdrawMessage.failure(err, meta))),
+  );
+}
+
+/**
+ * Handles a withdrawExpire.request, sign and send a WithdrawExpired message to partner
+ *
+ * @param state$ - Observable of current state
+ * @param action - WithdrawExpire request which caused this call
+ * @param deps - RaidenEpicDeps members
+ * @param deps.address - Our address
+ * @param deps.log - Logger instance
+ * @param deps.signer - Signer instance
+ * @param deps.config$ - Config observable
+ * @returns Observable of withdrawMessage.request|withdraw.failure actions
+ */
+function sendWithdrawExpired(
+  state$: Observable<RaidenState>,
+  action: withdrawExpire.request,
+  { log, address, signer, config$ }: RaidenEpicDeps,
+): Observable<withdrawExpire.success | withdrawExpire.failure | withdraw.failure> {
+  if (action.meta.direction !== Direction.SENT) return EMPTY;
+  return combineLatest([state$, config$]).pipe(
+    first(),
+    mergeMap(([state, { confirmationBlocks }]) => {
+      const channel = state.channels[channelKey(action.meta)];
+      assert(channel?.state === ChannelState.open, 'channel not open');
+
+      assert(
+        !channel.own.pendingWithdraws.some(
+          (exp) =>
```

Comment:
Yes, strictly you're right, this would be better as `msg` or something like that, it's just that the first condition filters it's an `Expired` message, so I already call it `exp`, to be more obvious in the following conditions. I'll rename, thanks.